### PR TITLE
Implement tables and call_indirect

### PR DESCRIPTION
**Pull request by pepyakin**
Based on #9 
Consists of only https://github.com/CraneStation/wasmtime/commit/de6b9f135a0df2a971ce332483316d0a4542334e for the moment of writing

### CODE REVIEW

File: lib/environ/src/environ.rs [hunk 1]
```diff
@@ -341,17 +359,31 @@ impl<'module_environment> cranelift_wasm::FuncEnvironment for FuncEnvironment<'m
...
-        // TODO: Cranelift's call_indirect doesn't implement bounds checking
-        // or signature checking, so we need to implement it ourselves.
         debug_assert_eq!(table_index, 0, "non-default tables not supported yet");
+
+        let callee_ty = pos.func.dfg.value_type(callee);
+        debug_assert_eq!(callee_ty, I32);
+
+        // For some reason table_addr only works with I64.
+        let callee_i64 = pos.ins().uextend(I64, callee);
+        let table_entry_addr = pos.ins().table_addr(I64, table, callee_i64, 0);
```
Comment (sunfishcode) [inline lib/environ/src/environ.rs]: The current limitation is that it be the same type as a pointer, fwiw.

File: lib/environ/src/environ.rs [hunk 2]
```diff
@@ -341,17 +359,31 @@ impl<'module_environment> cranelift_wasm::FuncEnvironment for FuncEnvironment<'m
...
+        table: ir::Table,
         _sig_index: SignatureIndex,
         sig_ref: ir::SigRef,
         callee: ir::Value,
         call_args: &[ir::Value],
     ) -> WasmResult<ir::Inst> {
-        // TODO: Cranelift's call_indirect doesn't implement bounds checking
-        // or signature checking, so we need to implement it ourselves.
         debug_assert_eq!(table_index, 0, "non-default tables not supported yet");
+
```
Comment (sunfishcode) [inline lib/environ/src/environ.rs]: We should leave the comment about signature checking for now.

File: lib/environ/src/environ.rs (hunk 1 above)
Comment (pepyakin) [reply lib/environ/src/environ.rs]: Oh, didn't know about that.

Then I would probably go with `pointer_type()` and extend only if needed.

File: lib/environ/src/environ.rs (hunk 2 above)
Comment (pepyakin) [reply lib/environ/src/environ.rs]: 😨 oh yeah you're right! That was unintended.

**Pull request merged**